File: app/navaids_db.py
"""Navaid identifier → coordinates + type.

Reads the OurAirports navaids CSV (public domain). Data at
https://github.com/davidmegginson/ourairports-data — parsed once at startup
and kept in memory. Drives the Navaids map overlay via the /api/navaids
bbox endpoint.

The file is baked into the Docker image at build time. If a user-provided
copy exists at /data/navaids.csv it wins (same override pattern as
airports.csv), otherwise we fall back to the baked-in version.
"""

import csv
import logging
from pathlib import Path
from typing import Any

log = logging.getLogger("beast.navaids_db")

DEFAULT_PATHS: list[Path] = [
    Path("/data/navaids.csv"),
    Path(__file__).parent / "navaids.csv",
]

# Types worth surfacing on the map. ILS glideslope / localizer / marker
# beacons aren't useful as independent overlay dots — they're always
# colocated with a runway threshold and would just clutter.
KEEP_TYPES = frozenset({"VOR", "VOR-DME", "VORTAC", "NDB", "NDB-DME", "DME", "TACAN"})

# Sort priority for bbox-culled responses — omnidirectional nav aids
# (VOR family) rank highest, DME-only and TACAN next, NDBs last. A wide
# view truncated to `limit` rows still includes the backbone of the
# airway network.
TYPE_RANK = {
    "VORTAC": 0,
    "VOR-DME": 0,
    "VOR": 1,
    "DME": 2,
    "TACAN": 2,
    "NDB-DME": 3,
    "NDB": 3,
}


class NavaidsDB:
    """In-memory ident → {name, type, lat, lon, frequency_khz} lookup."""

    def __init__(self) -> None:
        self._rows: list[dict[str, Any]] = []

    def __len__(self) -> int:
        return len(self._rows)

    def load_from(self, path: Path) -> int:
        """Parse OurAirports navaids CSV into a fresh list and swap in atomically."""
        new_rows: list[dict[str, Any]] = []
        with path.open("rt", encoding="utf-8", newline="") as fh:
            reader = csv.DictReader(fh)
            for row in reader:
                nav_type = (row.get("type") or "").strip().upper()
                if nav_type not in KEEP_TYPES:
                    continue
                ident = (row.get("ident") or "").strip().upper()
                name = (row.get("name") or "").strip()
                if not ident:
                    continue
                try:
                    lat = float(row.get("latitude_deg") or "")
                    lon = float(row.get("longitude_deg") or "")
                except ValueError:
                    continue
                freq_raw = (row.get("frequency_khz") or "").strip()
                try:
                    freq_khz: float | None = float(freq_raw) if freq_raw else None
                except ValueError:
                    freq_khz = None
                new_rows.append(
                    {
                        "ident": ident,
                        "name": name,
                        "type": nav_type,
                        "lat": lat,
                        "lon": lon,
                        "frequency_khz": freq_khz,
                        "country": (row.get("iso_country") or "").strip(),
                        "associated_airport": (row.get("associated_airport") or "").strip(),
                    }
                )
        self._rows = new_rows
        return len(self._rows)

    def bbox(
        self,
        min_lat: float,
        min_lon: float,
        max_lat: float,
        max_lon: float,
        limit: int = 2000,
    ) -> list[dict]:
        """Return navaids inside the bounding box, strongest-rank first.

        Handles antimeridian wrap when `min_lon > max_lon`. Truncates to
        `limit` rows sorted by type (VOR family → DME/TACAN → NDB).
        """
        wraps = min_lon > max_lon
        hits: list[dict] = []
        for r in self._rows:
            lat = r["lat"]
            lon = r["lon"]
            if not (min_lat <= lat <= max_lat):
                continue
            in_lon = (lon >= min_lon or lon <= max_lon) if wraps else (min_lon <= lon <= max_lon)
            if not in_lon:
                continue
            hits.append(r)
        hits.sort(key=lambda e: TYPE_RANK.get(e["type"], 9))
        return hits[:limit]

    def load_first_available(self, paths: list[Path] | None = None) -> int:
        for p in paths or DEFAULT_PATHS:
            if p.exists():
                try:
                    n = self.load_from(p)
                    log.info("loaded navaids DB from %s (%d entries)", p, n)
                    return n
                except Exception as e:
                    log.warning("failed to load navaids DB from %s: %s", p, e)
        log.info("no navaids DB found; overlay disabled")
        return 0

File: dotnet/src/FlightJar.Core/ReferenceData/NavaidsDb.cs
using System.Collections.Frozen;
using System.Globalization;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace FlightJar.Core.ReferenceData;

/// <summary>
/// Navaid identifier → coordinates + type. Reads the
/// <a href="https://github.com/davidmegginson/ourairports-data">OurAirports</a>
/// navaids CSV. Ports <c>app/navaids_db.py</c>.
/// </summary>
public sealed class NavaidsDb
{
    private static readonly FrozenSet<string> KeepTypes = new[]
    {
        "VOR", "VOR-DME", "VORTAC", "NDB", "NDB-DME", "DME", "TACAN",
    }.ToFrozenSet();

    private static readonly IReadOnlyDictionary<string, int> TypeRank = new Dictionary<string, int>
    {
        ["VORTAC"] = 0,
        ["VOR-DME"] = 0,
        ["VOR"] = 1,
        ["DME"] = 2,
        ["TACAN"] = 2,
        ["NDB-DME"] = 3,
        ["NDB"] = 3,
    };

    private readonly ILogger _logger;
    private List<NavaidRecord> _rows = new();

    public NavaidsDb(ILogger<NavaidsDb>? logger = null)
    {
        _logger = logger ?? NullLogger<NavaidsDb>.Instance;
    }

    public int Count => _rows.Count;

    public async Task<int> LoadFromAsync(string path, CancellationToken ct = default)
    {
        using var reader = File.OpenText(path);
        var fresh = new List<NavaidRecord>();
        await foreach (var row in CsvReader.ReadDictAsync(reader, ',', ct))
        {
            var type = Trim(row, "type").ToUpperInvariant();
            if (!KeepTypes.Contains(type))
            {
                continue;
            }
            var ident = Trim(row, "ident").ToUpperInvariant();
            if (ident.Length == 0)
            {
                continue;
            }
            if (!double.TryParse(Trim(row, "latitude_deg"), NumberStyles.Float, CultureInfo.InvariantCulture, out var lat)
                || !double.TryParse(Trim(row, "longitude_deg"), NumberStyles.Float, CultureInfo.InvariantCulture, out var lon))
            {
                continue;
            }
            double? freqKhz = null;
            var freqRaw = Trim(row, "frequency_khz");
            if (freqRaw.Length > 0
                && double.TryParse(freqRaw, NumberStyles.Float, CultureInfo.InvariantCulture, out var freq))
            {
                freqKhz = freq;
            }
            fresh.Add(new NavaidRecord(
                Ident: ident,
                Name: Trim(row, "name"),
                Type: type,
                Lat: lat,
                Lon: lon,
                FrequencyKhz: freqKhz,
                Country: Trim(row, "iso_country"),
                AssociatedAirport: Trim(row, "associated_airport")));
        }
        _rows = fresh;
        return _rows.Count;
    }

    public List<NavaidRecord> Bbox(
        double minLat, double minLon, double maxLat, double maxLon, int limit = 2000)
    {
        var wraps = minLon > maxLon;
        var hits = new List<NavaidRecord>();
        foreach (var r in _rows)
        {
            if (r.Lat < minLat || r.Lat > maxLat)
            {
                continue;
            }
            var inLon = wraps ? (r.Lon >= minLon || r.Lon <= maxLon) : (r.Lon >= minLon && r.Lon <= maxLon);
            if (!inLon)
            {
                continue;
            }
            hits.Add(r);
        }
        hits.Sort((a, b) =>
        {
            var ra = TypeRank.TryGetValue(a.Type, out var x) ? x : 9;
            var rb = TypeRank.TryGetValue(b.Type, out var y) ? y : 9;
            return ra.CompareTo(rb);
        });
        if (hits.Count > limit)
        {
            hits.RemoveRange(limit, hits.Count - limit);
        }
        return hits;
    }

    public async Task<int> LoadFirstAvailableAsync(IEnumerable<string> paths, CancellationToken ct = default)
    {
        foreach (var p in paths)
        {
            if (!File.Exists(p))
            {
                continue;
            }
            try
            {
                var n = await LoadFromAsync(p, ct);
                _logger.LogInformation("loaded navaids DB from {Path} ({Count} entries)", p, n);
                return n;
            }
            catch (Exception ex) when (ex is not OperationCanceledException)
            {
                _logger.LogWarning(ex, "failed to load navaids DB from {Path}", p);
            }
        }
        _logger.LogInformation("no navaids DB found; overlay disabled");
        return 0;
    }

    private static string Trim(Dictionary<string, string> row, string key)
        => row.TryGetValue(key, out var v) ? v.Trim() : "";
}

public sealed record NavaidRecord(
    string Ident,
    string Name,
    string Type,
    double Lat,
    double Lon,
    double? FrequencyKhz,
    string Country,
    string AssociatedAirport);
